File: Day-01/python/Schloool/day_01.py
input = open('input.txt').readlines()

letters = {
    'one': 1,
    'two': 2,
    'three': 3,
    'four': 4,
    'five': 5,
    'six': 6,
    'seven': 7,
    'eight': 8,
    'nine': 9
}


def get_max_number_from_index(line, index, account_number_words):
    if line[index].isnumeric():
        return line[index]

    if not account_number_words:
        return None

    for key in reversed(letters.keys()):
        if len(key) > len(line) - index:
            continue
        if line[index:(index + len(key))] == key:
            return letters[key]
    return None


def get_num_pair(account_number_words):
    nums = []
    for line in input:
        first, last = None, None
        i = 0

        while first is None:
            first = get_max_number_from_index(line, i, account_number_words)
            i += 1 if first is None else 0

        for j in range(i, len(line)):
            found_num = get_max_number_from_index(line, j, account_number_words)
            last = str(found_num) if found_num is not None else last

        nums.append(int(str(first) + last))
    return nums


# Task 1
print(sum(get_num_pair(False)))

# Task 2
print(sum(get_num_pair(True)))

File: Day-01/rust/emmaboecker/day01.rs
const WRITTEN_NUMBERS: [&str; 9] = [
    "one", "two", "three", "four", "five", "six", "seven", "eight", "nine",
];

pub fn main() {
    let input = include_str!("../../input/day01.txt");
    println!("Part 1: {}", find_numbers(input, false));
    println!("Part 2: {}", find_numbers(input, true));
}

fn find_numbers(input: &str, written: bool) -> u32 {
    input
        .lines()
        .map(|line| {
            let valid = find_valid(line, written);
            format!("{}{}", valid.first().unwrap(), valid.last().unwrap())
                .parse()
                .unwrap()
        })
        .collect::<Vec<u32>>()
        .iter()
        .sum()
}

fn find_valid(input: &str, written: bool) -> Vec<u32> {
    let mut valid = vec![];

    if written {
        for (index, written) in WRITTEN_NUMBERS.iter().enumerate() {
            let indices = input.match_indices(written);

            for (position, _) in indices {
                valid.push((position, index + 1));
            }
        }
    }

    for value in 1..10 {
        let indices = input.match_indices((value as u8 + b'0') as char);

        for (index, _) in indices {
            valid.push((index, value));
        }
    }

    valid.sort_by(|a, b| a.0.cmp(&b.0));
    valid
        .into_iter()
        .map(|(_, value)| value as u32)
        .collect::<Vec<u32>>()
}